main.nf

```diff
@@ -140,392 +165,666 @@ process fastqc {
 
+      gzip \${prefix}.mature_unmapped.fq
+      """
+    } else {
+      """
+      f='$reads';f=(\$f);input=\${f[0]};f=\${input%.gz};f=\${f%.fastq};f=\${f%.fq};f=\${f%.R1_val_1*};f=\${f%.R2_val_2*};f=\${f%_val_1};f=\${f%_val_2};f=\${f%_trimmed};f=\${f%_1}
+      prefix=\$f
+
+      bowtie $mature -q <(zcat \$input) -v 0 -k 10 -S \${prefix}.mature.sam --un \${prefix}.mature_unmapped.fq
+
+      gzip \${prefix}.mature_unmapped.fq
+      """
+    }
+}
 
 /*
- * STEP 4 - Bowtie against miRBase hairpin RNA
+ * STEP 3.2 - Bowtie against miRBase mature miRNA post-alignment processing
  */
-process bowtie_miRBase_hairpin {
-    
+
+process bowtie_miRBase_mature_postalignment {
+
+    tag "$miRBase_mature_sam"
+
     module 'bioinfo-tools'
-    module 'bowtie'
     module 'samtools'
-    
-    memory '2 GB'
-    time '6h'
+    module 'BEDTools'
+
+    cpus 2
+    memory { 16.GB * task.attempt }
+    time { 120.h * task.attempt }
+    errorStrategy { task.exitStatus == 143 ? 'retry' : 'terminate' }
+    maxRetries 3
+    maxErrors '-1'
+
+    publishDir "${params.outdir}/bowtie_miRBase_mature", mode: 'copy'
 
     input:
-    file(reads:'*') from trimmed_reads
-    
+    file miRBase_mature_sam
+
     output:
-    file '*.bam' into miRBase_hairpin_bam
-    
+    file '*.sorted.bam' into miRBase_mature_bam
+    file '*.sorted.bam.bai' into miRBase_mature_bai
+    file '*.count' into miRBase_mature_count
+
     script:
     """
-    f='$reads';f=(\$f);f=\${f[0]};f=\${f%.gz};f=\${f%.fastq};f=\${f%.fq};f=\${f%_val_1};f=\${f%_trimmed};f=\${f%_1}
-    bowtie \
-        -p 2 \
-        -t \
-        -n 0 \
-        -l 15 \
-        -e 99999 \
-        -k 200 \
-        --best \
-        -S \
-        --chunkmbs 2048 \
-        $bowtie_miRBase_hairpin \
-        $reads \
-        | samtools view -bS > \${f}.aligned_miRBase_hairpin.bam
+    f='$miRBase_mature_sam';f=(\$f);f=\${f[0]};f=\${f%.sam}
+    prefix=\$f
+
+    samtools view -bS \${prefix}.sam > \${prefix}.bam
+    samtools sort \${prefix}.bam \${prefix}.sorted
+    samtools index \${prefix}.sorted.bam
+    samtools idxstats \${prefix}.sorted.bam > \${prefix}.count
+
+    rm \${prefix}.sam
+    rm \${prefix}.bam
     """
 }
 
 
 /*
- * STEP 5 - Samtools idxstats to count bowtie 1 alignments
+ * STEP 3.3 - edgeR miRBase mature miRNA counts processing
  */
-process samtools_idxstats {
-    
+
+process edgeR_miRBase_mature {
+
     module 'bioinfo-tools'
-    module 'samtools'
-    
-    memory '8 GB'
-    time '6h'
+    module 'R/3.2.3'
+
+    cpus 2
+    memory { 16.GB * task.attempt }
+    time { 120.h * task.attempt }
+    errorStrategy { task.exitStatus == 143 ? 'retry' : 'terminate' }
+    maxRetries 3
+    maxErrors '-1'
+
+    publishDir "${params.outdir}/bowtie_miRBase_mature/edgeR", mode: 'copy'
 
     input:
-    file '*.bam' from miRBase_mature_bam
-    // TODO: WORK OUT HOW TO GET BOTH BOWTIE CHANNELS TO COME HERE
-    // ADD: miRBase_hairpin_bam
```

Comment:
I'd prefer to have a single process do this for both mature and hairpin results if possible. I don't like duplicating the code if we can avoid it.